libraries/ESP8266WebServer/examples/FSBrowser/readme.md

```diff
@@ -0,0 +1,121 @@
+# FSBrowser readme
+
+## What is this sketch about ?
+
+This example is a FileSystem Browser for the ESP8266 using http requests and a html/javascript frontend, 
+working for both SPIFFS, LittleFS and SDFS.
+This unified version is based on the previous examples named FSWebServer, FSBrowser and SDWebServer, Copyright (c) 2015 Hristo Gochkov. All rights reserved.
+
+## How to use ?
+1. Uncomment one of the `#define USE_xxx` directives in the sketch
+2. Add the credentials of your WiFi network (search for `STASSID`)
+3. Compile and upload the sketch to your ESP8266 device
+4. For normal use, copy the contents of the `data` folder to the filesystem. To do so:
+- for SDFS, copy that contents (not the data folder itself, just its contents) to the root of a FAT/FAT32-formated SD card connected to the SPI port of the ESP8266
+- for SPIFFS or LittleFS, please follow the instructions at https://arduino-esp8266.readthedocs.io/en/latest/filesystem.html#uploading-files-to-file-system
+5. Once the data and sketch have been uploaded, access the editor by pointing your browser to http://fsbrowser.local/edit
+
+## Options
+If you need to free some space on the ESP filesystem, you can delete all the sample files at the root but also replace the `index.htm` file in the `data/edit` subfolder by the `index.htm.gz` file from the `extras` folder. That compressed version is not suited for learning or debugging, but will bring the total FS usage under 7KB.
+If you want to use the browser on a an existing filesystem or don't want to perform step 4 above, you have two possibilities :
+- either upload the `index.htm` file to the filesystem by opening a command shell in the `data` folder and running the following cURL command:
+`curl -F file=@edit/index.htm;filename=/edit/index.htm fsbrowser.local/edit`
+- or embed a version of the html page in the source code itself by uncommenting the following line in the sketch and rebuilding:
+`#define INCLUDE_FALLBACK_INDEX_HTM`
+That embedded version is functionally equivalent and will be returned if no `/edit/index.htm` or `/edit/index.htm.gz` file can be found on the filesystem, at the cost of a larger compiled file.
+
+To regenerate the files required for the previous options or understand how they are build, see the `reduce_index.sh` script located in the `extras` subfolder.
+
+## Dependency
+The html page uses the [ace.js](https://ace.c9.io/) text editor which is loaded from a CDN. Consequently, internet access from your web browser is required for the FSBrowser editing feature to work as-is.
+If your browser has no web access (e.g. if you are connected to the ESP8266 as an access-point), you can copy the `ace.js` file (and dependencies) to the `edit` subfolder of the ESP filesystem.
```

Comment:
How do I figure out the dependencies for ace? It's not a matter of going in there and poking around.

Reply:
Hmm, I admit I'm a bit puzzled too, and can't find a definitive answer. 
Seems ace.js dynamically loads additional files according to supported file formats, preferred keymaps, themes, etc. The full list is at https://github.com/ajaxorg/ace-builds/tree/master/src but I believe the editor works with none loaded and just falls back to default settings if files cannot be loaded. 
I'd suggest including this in the readme :
`copy the `ace.js` file to the `edit` subfolder of the ESP filesystem, as well as optional themes, plugins etc. according to your needs (see https://github.com/ajaxorg/ace-builds for a full list).`
What do you think ?

Reply:
Agreed. However, the most widely used cases for this editor will be html, js, css and json. What do you think of figuring out the minimum files for those specific formats and suggesting those explicitly together with the above comment? I'm trying to think like the newcomer looking at the example for the first time. Looking at that full list is rather overwhelming.
So the comment would be something like:
```
Copy the ace.js file to the edit subfolder of the ESP filesystem, as well as optional themes, plugins etc. according to your needs (see https://github.com/ajaxorg/ace-builds for a full list).
Example:
mode-css.js
mode-html.js
blah.js
...
```

Reply:
Indeed, I admit as a newcomer to ace.js, I find the list overwhelming ;-).
OK, I'll try to figure out a "reasonable" list and will update the readme accordingly. For example, the theme is fixed in index.htm IIRC, so it is easy to determine which theme file is needed.

Reply:
Done. 
I just opened the "Network" developer tool and played with the editori to collect what was actually loaded from the CDN and compile a "typical" list :-).